In [1]:
import empylib.waveoptics as wv
import empylib.rad_transfer as rt


cases = {'case1': {'N_back': 10.0, 'nh': 1.0, 'tfilm_test': 5.0, 'Dp': 0.3, 'fv': 0.0},             # air/material with no particles
         'case2': {'N_back': 10.0, 'nh': 1.0, 'tfilm_test': 5.0, 'Dp': 0.3, 'fv': 0.5},             # air/material with 50% air particles
         'case3': {'N_back': 10.0, 'nh': 1.0, 'tfilm_test': 0, 'Dp': 0.3, 'fv': 0.5},               # air/material with no thickness
         'case4': {'N_back': 10.0+20j, 'nh': 1.0, 'tfilm_test': 5.0, 'Dp': 0.3, 'fv': 0.0},         # air/metal with no particles
         'case5': {'N_back': 10.0+20j, 'nh': 1.0, 'tfilm_test': 5.0, 'Dp': 0.3, 'fv': 0.5},         # air/metal with 50% air particles
         'case6': {'N_back': 10.0+20j, 'nh': 1.0, 'tfilm_test': 0, 'Dp': 0.3, 'fv': 0.5},           # air/metal with no thickness
         'case7': {'N_back': 10.0+50j, 'nh': 1.5+1E-5j, 'tfilm_test': 5.0, 'Dp': 0.3, 'fv': 0.0},    # absorptive/metal with no particles
         'case8': {'N_back': 10.0+50j, 'nh': 1.5+1E-5j, 'tfilm_test': 5.0, 'Dp': 0.3, 'fv': 0.5},    # absorptive/metal with 50% absorptive particles
         'case9': {'N_back': 10.0+50j, 'nh': 1.5+1E-5j, 'tfilm_test': 0, 'Dp': 0.3, 'fv': 0.5},      # absorptive/metal with no thickness
         'case10': {'N_back': 1.0, 'nh': 1.5+1E-5j, 'tfilm_test': 5.0, 'Dp': 0.3, 'fv': 0.0},    # absorptive/metal with no particles
         'case11': {'N_back': 1.0, 'nh': 1.5+1E-5j, 'tfilm_test': 5.0, 'Dp': 0.3, 'fv': 0.5},    # absorptive/metal with 50% absorptive particles
         'case12': {'N_back': 1.0, 'nh': 1.5+1E-5j, 'tfilm_test': 0, 'Dp': 0.3, 'fv': 0.5},      # absorptive/metal with no thickness
         'case13': {'N_above': 1.0, 'N_back': 1.0, 'nh': 1.0, 'tfilm_test': 5.0, 'Dp': 0.3, 'fv': 0.0},               # uniform non-absorbing medium, finite thickness
         'case14': {'N_above': 1.0, 'N_back': 1.5, 'nh': 1.5, 'tfilm_test': 5.0, 'Dp': 0.3, 'fv': 0.0},               # top dielectric interface only, finite thickness
         'case15': {'N_above': 1.0, 'N_back': 1.5, 'nh': 1.5, 'tfilm_test': 0.0, 'Dp': 0.3, 'fv': 0.5},               # top dielectric interface only, zero thickness
         'case16': {'N_above': 1.0, 'N_back': 1.0, 'nh': 1.5, 'tfilm_test': 5.0, 'Dp': 0.3, 'fv': 0.0},               # air/dielectric/air finite slab
         'case17': {'N_above': 1.0, 'N_back': 1.0, 'nh': 1.5, 'tfilm_test': 0.0, 'Dp': 0.3, 'fv': 0.5},               # air/dielectric/air zero thickness
         'case18': {'N_above': 1.5+1E-5j, 'N_back': 1.5+1E-5j, 'nh': 1.5+1E-5j, 'tfilm_test': 5.0, 'Dp': 0.3, 'fv': 0.0},  # uniform absorptive medium, finite thickness
         'case19': {'N_above': 1.0, 'N_back': 1.5+1E-5j, 'nh': 1.5+1E-5j, 'tfilm_test': 5.0, 'Dp': 0.3, 'fv': 0.0},  # top absorptive interface only, finite thickness
         'case20': {'N_above': 1.0, 'N_back': 1.5+1E-5j, 'nh': 1.5+1E-5j, 'tfilm_test': 0.0, 'Dp': 0.3, 'fv': 0.5},  # top absorptive interface only, zero thickness
         'case21': {'N_above': 1.0, 'N_back': 1.0, 'nh': 1.5+1E-5j, 'tfilm_test': 5.0, 'Dp': 0.3, 'fv': 0.0},         # air/absorptive/air finite slab
         'case22': {'N_above': 1.0, 'N_back': 1.0, 'nh': 1.5+1E-5j, 'tfilm_test': 0.0, 'Dp': 0.3, 'fv': 0.5},         # air/absorptive/air zero thickness
         }
lam_test = 0.5


print(f'|case | interface | incoh_multilayer | adm_sphere |')
print(f'|:---:|:---------:|:----------------:|:----------:|')
for case, values in cases.items():
    N_back = values['N_back']
    nh = values['nh']
    N_above = values.get('N_above', nh)
    tfilm_test = values['tfilm_test']
    Dp = values['Dp']
    fv = values['fv']

    Ri, Ti = wv.incoh_multilayer(wavelength = lam_test,
                                 N_layers = nh,
                                 N_above = N_above,
                                 N_below = N_back,
                                 thickness = tfilm_test*1E3)

    Rj, Tj, *_ = wv.interface(N_above, N_back)

    out_test = rt.adm_sphere(wavelength = lam_test,
                           N_host = nh,
                           N_particle = nh,
                           D = Dp,
                           fv = fv,
                           thickness = tfilm_test,
                           N_above = N_above,
                           N_below = N_back,
                           effective_medium = True,
                           use_phase_fun = True,
                           dependent_scatt = True)
    # print(Ri, Rj, out_test.Rtot[0])
    print(f'|{case} | {Rj[0]:.6f} | {Ri[0]:.6f} | {out_test.Rtot.values[0]:.6f} |')

|case | interface | incoh_multilayer | adm_sphere |
|:---:|:---------:|:----------------:|:----------:|
|case1 | 0.669421 | 0.669421 | 0.801980 |
|case2 | 0.669421 | 0.669421 | 0.801980 |
|case3 | 0.669421 | 0.669421 | 0.801980 |
|case4 | 0.923225 | 0.923225 | 0.801980 |
|case5 | 0.923225 | 0.923225 | 0.801980 |
|case6 | 0.923225 | 0.923225 | 0.801980 |
|case7 | 0.977205 | 0.079156 | 0.097140 |
|case8 | 0.977205 | 0.079156 | 0.056235 |
|case9 | 0.977205 | 0.977205 | 0.765854 |
|case10 | 0.040000 | 0.003240 | 0.042986 |
|case11 | 0.040000 | 0.003240 | 0.040850 |
|case12 | 0.040000 | 0.040000 | 0.076923 |
|case13 | 0.000000 | 0.000000 | 0.000000 |
|case14 | 0.040000 | 0.040000 | 0.076923 |
|case15 | 0.040000 | 0.040000 | 0.076923 |
|case16 | 0.000000 | 0.076923 | 0.076923 |
|case17 | 0.000000 | 0.000000 | 0.076923 |
|case18 | 0.000000 | 0.000000 | 0.042986 |
|case19 | 0.040000 | 0.040000 | 0.042986 |
|case20 | 0.040000 | 0.040000 | 0.076923 |
|case21 | 0.000000 | 0.042986 | 0.042986 |
|c

| case | interface | incoh_multilayer | adm_sphere | comments |
|:----:|:---------:|:----------------:|:----------:|:---------|
| case1 | 0.669421 | 0.669421 | 0.801980 | Physically right: `interface`, `incoh_multilayer`. Fails: `adm_sphere`. No scattering and the film is index-matched to the incident medium, so this should reduce to the bottom Fresnel interface. |
| case2 | 0.669421 | 0.669421 | 0.801980 | Physically right: `interface`, `incoh_multilayer`. Fails: `adm_sphere`. Since `N_particle = N_host`, even `fv = 0.5` should not change the optics. |
| case3 | 0.669421 | 0.669421 | 0.801980 | Physically right: `interface`, `incoh_multilayer`. Fails: `adm_sphere`. Zero thickness must collapse exactly to the interface limit. |
| case4 | 0.923225 | 0.923225 | 0.801980 | Physically right: `interface`, `incoh_multilayer`. Fails: `adm_sphere`. Same limiting case as above, but with a metallic substrate. |
| case5 | 0.923225 | 0.923225 | 0.801980 | Physically right: `interface`, `incoh_multilayer`. Fails: `adm_sphere`. The 50% volume fraction is optically irrelevant here because the particles are identical to the host. |
| case6 | 0.923225 | 0.923225 | 0.801980 | Physically right: `interface`, `incoh_multilayer`. Fails: `adm_sphere`. Zero-thickness limit should again be the bare interface. |
| case7 | 0.977205 | 0.079156 | 0.097140 | Physically right: `incoh_multilayer`. Fails: `adm_sphere`. `interface` is also not the right physics here because it ignores the finite, absorptive slab thickness. |
| case8 | 0.977205 | 0.079156 | 0.056235 | Physically right: `incoh_multilayer`. Fails: `adm_sphere`. `interface` is again not physically appropriate for a finite absorptive slab. |
| case9 | 0.977205 | 0.977205 | 0.765854 | Physically right: `interface`, `incoh_multilayer`. Fails: `adm_sphere`. With zero thickness, the absorptive slab disappears and the result must revert to the interface. |
| case10 | 0.040000 | 0.003240 | 0.042986 | Physically right: `incoh_multilayer`. Fails: `adm_sphere`. `interface` is also not the right model because it neglects absorption across the finite slab. |
| case11 | 0.040000 | 0.003240 | 0.040850 | Physically right: `incoh_multilayer`. Fails: `adm_sphere`. The finite absorptive slab should dominate; adding particles identical to the host should not change that limit. |
| case12 | 0.040000 | 0.040000 | 0.076923 | Physically right: `interface`, `incoh_multilayer`. Fails: `adm_sphere`. Zero thickness should reduce to the direct interface with no slab propagation. |
| case13 | 0.000000 | 0.000000 | 0.000000 | Physically right: `interface`, `incoh_multilayer`, `adm_sphere`. This is a clean control: a uniform non-absorbing medium with no interfaces should have zero reflectance, and all three models agree. |
| case14 | 0.040000 | 0.040000 | 0.076923 | Physically right: `interface`, `incoh_multilayer`. Fails: `adm_sphere`. With substrate matched to the host, only the top dielectric interface remains, so slab thickness should not matter. |
| case15 | 0.040000 | 0.040000 | 0.076923 | Physically right: `interface`, `incoh_multilayer`. Fails: `adm_sphere`. Zero thickness should still collapse to that same top-interface limit. |
| case16 | 0.000000 | 0.076923 | 0.076923 | Physically right: `incoh_multilayer`, `adm_sphere`. `interface` is not the right model here because the finite air/dielectric/air slab has two boundaries. |
| case17 | 0.000000 | 0.000000 | 0.076923 | Physically right: `interface`, `incoh_multilayer`. Fails: `adm_sphere`. Zero thickness removes the slab completely, so the air/air result must be zero reflectance. |
| case18 | 0.000000 | 0.000000 | 0.042986 | Physically right: `interface`, `incoh_multilayer`. Fails: `adm_sphere`. A uniform absorptive medium with no interfaces can absorb, but it still should not reflect. |
| case19 | 0.040000 | 0.040000 | 0.042986 | Physically right: `interface`, `incoh_multilayer`. `adm_sphere` is close but still fails. With substrate matched to the absorptive host, only the top interface should contribute to reflectance. |
| case20 | 0.040000 | 0.040000 | 0.076923 | Physically right: `interface`, `incoh_multilayer`. Fails: `adm_sphere`. Zero thickness should again collapse to the top-interface value. |
| case21 | 0.000000 | 0.042986 | 0.042986 | Physically right: `incoh_multilayer`, `adm_sphere`. `interface` is not the right model because the finite absorptive air/slab/air geometry needs slab propagation. |
| case22 | 0.000000 | 0.000000 | 0.076923 | Physically right: `interface`, `incoh_multilayer`. Fails: `adm_sphere`. Zero thickness should give the air/air limit, not a residual slab reflectance. |
